### ISSUE TITLE
onMessageReceivedCallback works in foreground and background but not if app is closed (e.g. swiped from recents)

### ISSUE THREAD

**Issue by trashbytes** [EddyVerbruggen/nativescript-plugin-firebase#632]
The onMessageReceivedCallback handler works just fine when the app is running in foreground or when I minimize it. Whenever I close it, for example by swiping it from the recents, it will not work anymore.

When it's in this state the only notifications that come through (to the tray) are these kind:

`notification: {
  title: 'Title',
  body: 'Body'
}`

But what I need is this: 

`data = {
  messageId: '1519656340496',
  userId: 'USER_ID',
  date: '1519656340496',
  text: 'Text Message',
  type: 'message',
  messageCollectionId: 'messages',
  channelId: 'CHANNEL_ID',
  displayName: 'USER_NAME'
}`

However the app only receives them after launching the app.

This line is already active in the include.gradle file:
`compile "com.google.firebase:firebase-messaging:$firebaseVersion"`

Both services are enabled in AppManifest.xml:
`<service android:name="org.nativescript.plugins.firebase.MyFirebaseInstanceIDService">
  <intent-filter>
    <action android:name="com.google.firebase.INSTANCE_ID_EVENT"/>
  </intent-filter>
</service>
<service android:name="org.nativescript.plugins.firebase.MyFirebaseMessagingService">
  <intent-filter>
    <action android:name="com.google.firebase.MESSAGING_EVENT"/>
  </intent-filter>
</service>`

onMessageReceivedCallback is initialized directly in the .init() function in app.js.
I'm sending to a topic which the user subscribes to when he logs in.

**Comment by trashbytes** [comment]
@EddyVerbruggen any idea?

**Comment by EddyVerbruggen** [comment]
Not really. Perhaps move that init logic out of app.js, and keep the require of the plugin in there.

**Comment by trashbytes** [comment]
you mean the whole firebase.init() out of app.js but still leaving require('nativescript-plugin-firebase')?

**Comment by EddyVerbruggen** [comment]
yes..

**Comment by trashbytes** [comment]
This does not work. Is it even remotely possible that onMessageReceivedCallback is called when the app is suspended? Or do I have to use 'notification' instead of 'data' if I want to receive notifications in that state?

**Comment by erkanarslan** [comment]
@trashbytes I searched about that problem and found that firebase won't notify your app for a new notification if user closed the app by swiping. They say that user closes the app consciously if he is swiping it and they respect that. So, they don't notify your app. I think there is nothing else to do about that.

**Comment by mayureshjadhav** [comment]
+1

**Comment by KkevinLi** [comment]
@erkanarslan I do not believe this is true. Using the Quickstart Android demo (https://github.com/firebase/quickstart-android/blob/master/messaging/README.md) I am able to get onMessageReceived() even after swiping the app and from there I can build local notifications to show the user. 

On the other hand, even with the latest 6.20 nativescript plugin, onMessageReceived() will not be called for me once I swipe the app. My manifest is identical to the one from Android Quickstart and get the same behavior as @trashbytes.

**Comment by trashbytes** [comment]
Thanks for sharing @KkevinLi ! Please keep us updated if you make any advancements regarding this. For us it's not currently a priority but still on the list of things which we want to get working so that we can achieve richer notifications.

**Comment by felix-idf** [comment]
I have the same issue with the latest version of this plugin (7.1.2) when launchMode is set to **singleTask**.
The problem is not only that the callback is not onMessageReceivedCallback , I don't see any notification at all (there is nothing in the notification tray).

Before, I used the nativescript-push-notifications plugin and this issue did not occur with that plugin.
But since it is marked as deprecated, I switched to this nativescript-plugin-firebase.

Setting the launchMode to **singleTop** works for the push messages when the app is completely closed, but then I have troubles when handling links, as the app seems to be reopened always when I tap on a link and call my handler.

Setting the launchMode to **singleInstance** seems to do the trick here, as I dont see issues with my link handler and with the push notifications. But I am not sure if this creates any side effects in other situations and it seems to be not recommended on several Android docs.

In addition I need to say, that I did not add any firebase services to the AndroidManifest like other users mentioned in this repository, because it is not mentioned in the docs of this plugin.

**Comment by caiusCitiriga** [comment]
@KkevinLi any news on this?

**Comment by marianadgeorgieva2** [comment]
Hi. Is there any update on this?

**Comment by dangersvn** [comment]
Hi @EddyVerbruggen ,
I got the same issue. It is working perfectly when the app in Background or in Foreground but when the app exit then i receive nothing.

Thank you!

**Comment by erkanarslan** [comment]
@dangersvn If you are sending a data message and the application is closed, you won't get a notification in notification tray. Try sending a notification message. If you are already sending a notification message, some devices seems to be preventing FCM messages to preserve the battery. Try with another device. If you are trying with an iPhone simulator, as far as I know you cannot get a notification for ios simulators.

**Comment by dangersvn** [comment]
Hi @erkanarslan 
Thank you very much for your feedback!
I tried with samsung s8 device + firebase + postman to verify the different between the Data message and Notification message and exactly as you described. I have not tested with iOS simulator yet so no feedback.
In my case i have to deal with Data message, since the app depends on a third party backend which always using Data message to send PN via firebase to users while he/she offline. 
In case can't modify the third party backend then i need to use a service that is running on background even when the app is closed to deal with Data message.

**Comment by hkurma** [comment]
@erkanarslan When my app is closed and I send a notification message I do receive the notification in the notification tray, but the onMessageReceivedCallback() is not called (I know this since my console.log statement did log anything). Is it the desired behaviour ?

**Comment by erkanarslan** [comment]
@hkurma This the normal behavior. When you tap the notification, onMessageReceivedCallback() should be called.

**Comment by gustavost26** [comment]
Still don't have any concrete solution to this problem?
